src/System.Collections/src/System/Collections/Generic/HashSet.cs

```diff
@@ -299,14 +355,15 @@ public bool Remove(T item)
         {
             if (_buckets != null)
             {
-                int hashCode = InternalGetHashCode(item);
+                IEqualityComparer<T> comparer = _comparer ?? EqualityComparer<T>.Default;
```

Comment:
Do these null checks introduced, impact perf?

Reply:
I just realize that the code here may not be optimal.

In `Dictionary<T>` it uses code like below to get hash code and compare two keys for the case without using `EqualityComparer<TKey>.Default` intrinsic (e.g. Remove):

```csharp
// get hash code
comparer?.GetHashCode(key) ?? key.GetHashCode()
// compare keys
comparer?.Equals(entry.key, key) ?? EqualityComparer<TKey>.Default.Equals(entry.key, key)
```

It seems like having saved one more indirection by checking if the `comparer` is null, but it also requires null check for each call. I'll do a quick benchmark to confirm this micro-optimization and submit another commit.

Reply:
The benchmark calls `GetHashCode` once and `Equals` twice to simulate a lookup in methods like `Remove`.

```csharp
public abstract class DefaultComparerPerf<T>
{
    private IEqualityComparer<T> _comparer;
    public T Value1;
    public T Value2;

    public T Key;

    [Benchmark]
    public int AssignDefault()
    {
        var comparer = _comparer ?? EqualityComparer<T>.Default;
        var hashCode = comparer.GetHashCode(Key);
        var b1 = comparer.Equals(Value1, Key);
        var b2 = comparer.Equals(Value2, Key);
        return hashCode;
    }

    [Benchmark]
    public int NullCheck()
    {
        var comparer = _comparer;
        var hashCode = comparer?.GetHashCode(Key) ?? Key.GetHashCode();
        var b1 = comparer?.Equals(Value1, Key) ?? EqualityComparer<T>.Default.Equals(Value1, Key);
        var b2 = comparer?.Equals(Value1, Key) ?? EqualityComparer<T>.Default.Equals(Value1, Key);
        return hashCode;
    }
}

public class DefaultComparerPerf_Int32 : DefaultComparerPerf<int> { }
public class DefaultComparerPerf_Object : DefaultComparerPerf<object>
{ 
    public DefaultComparerPerf_Object()
    {
        Key = new object();
    }
}
```

Benchmark shows for `int` it takes 75% less time, but for `object` it takes 40% more:

``` ini
BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17763.805 (1809/October2018Update/Redstone5)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100
  [Host]     : .NET Core 3.0.0 (CoreCLR 4.700.19.46205, CoreFX 4.700.19.46214), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0 (CoreCLR 4.700.19.46205, CoreFX 4.700.19.46214), 64bit RyuJIT
```
Result for `int`:

|        Method |     Mean |     Error |    StdDev |
|-------------- |---------:|----------:|----------:|
| AssignDefault | 6.177 ns | 0.0693 ns | 0.0614 ns |
|     NullCheck | 1.427 ns | 0.0551 ns | 0.0516 ns |

Result for `object`:

|        Method |     Mean |     Error |    StdDev |
|-------------- |---------:|----------:|----------:|
| AssignDefault | 23.43 ns | 0.1344 ns | 0.1122 ns |
|     NullCheck | 33.99 ns | 0.7669 ns | 1.3228 ns |

In theory we can do the same thing for `Remove` as in lookup but that's what `Dictionary<,>` currently doing so I just keep it the same.